src/lib/ledger_builder/ledger_builder.ml

```diff
@@ -302,6 +302,78 @@ end = struct
 
     let hash t = Ledger_builder_aux_hash.of_bytes (hash_to_string t)
 
+    let create_expected_statement
+        {Super_transaction_with_witness.transaction_with_info; witness; _} =
+      let open Or_error.Let_syntax in
+      let source =
+        Frozen_ledger_hash.of_ledger_hash @@ Sparse_ledger.merkle_root witness
+      in
+      let%bind transaction =
+        Ledger.Undo.super_transaction transaction_with_info
+      in
+      let%bind after =
+        Or_error.try_with (fun () ->
+            Sparse_ledger.apply_super_transaction_exn witness transaction )
+      in
+      let target =
+        Frozen_ledger_hash.of_ledger_hash @@ Sparse_ledger.merkle_root after
+      in
+      let%bind fee_excess = Super_transaction.fee_excess transaction in
+      let%map supply_increase =
+        Super_transaction.supply_increase transaction
+      in
+      { Ledger_proof_statement.source
+      ; target
+      ; fee_excess
+      ; supply_increase
+      ; proof_type= `Base }
+
+    let scan_statement t :
+        (Ledger_proof_statement.t, [`Error of Error.t | `Empty]) Result.t =
+      let open Container.Continue_or_stop in
+      let stop e = Stop (Error e) in
+      let merge_acc (acc: Ledger_proof_statement.t option) s2 =
+        match acc with
+        | None -> Continue (Some s2)
+        | Some s1 ->
+          match Ledger_proof_statement.merge s1 s2 with
+          | Ok statement -> Continue (Some statement)
+          | Error e -> stop e
+      in
+      let fold_step acc_statement job =
+        match job with
+        | Parallel_scan.State.Job.Merge (None, Some (p, _))
+         |Merge (Some (p, _), None) ->
+            merge_acc acc_statement (Ledger_proof.statement p)
+        | Merge (None, None) -> Continue acc_statement
+        | Merge (Some (proof_1, _), Some (proof_2, _)) -> (
+          match
+            Ledger_proof_statement.merge
+              (Ledger_proof.statement proof_1)
+              (Ledger_proof.statement proof_2)
+          with
+          | Ok merged_statement -> merge_acc acc_statement merged_statement
+          | Error e -> stop e )
+        | Base None -> Continue acc_statement
+        | Base (Some transaction) ->
+          match create_expected_statement transaction with
+          | Ok expected_statement ->
+              if
+                Ledger_proof_statement.equal transaction.statement
+                  expected_statement
+              then merge_acc acc_statement transaction.statement
+              else stop (Error.of_string "Bad base statement")
+          | Error e -> stop e
+      in
+      let res =
+        Parallel_scan.State.fold_chronological_until t ~init:None
+          ~finish:Or_error.return ~f:fold_step
+      in
+      match res with
+      | Ok None -> Error `Empty
+      | Ok (Some res) -> Ok res
+      | Error e -> Error (`Error e)
+
```

Comment:
As discussed, `scan_statement` and `scan_statement_and_verify` can be one function with `verify_proof` trivially true for the scan_statement case